hudi-client/hudi-spark-client/src/test/java/org/apache/hudi/client/TestHoodieClientMultiWriter.java

```diff
@@ -545,6 +546,116 @@ public void testMultiWriterWithAsyncTableServicesWithConflict(HoodieTableType ta
+                    .build()).withAutoCommit(false).withProperties(lockProperties);
+    Set<String> validInstants = new HashSet<>();
+    // Create the first commit with inserts
+    HoodieWriteConfig cfg = writeConfigBuilder.build();
+    SparkRDDWriteClient client = getHoodieWriteClient(cfg);
+    createCommitWithInserts(cfg, client, "000", "001", 200, true);
+    validInstants.add("001");
+
+    // Three clients running actions in parallel
+    final int threadCount = 3;
+    final ExecutorService executor = Executors.newFixedThreadPool(threadCount);
+
+    final SparkRDDWriteClient client1 = getHoodieWriteClient(cfg);
+    final SparkRDDWriteClient client2 = getHoodieWriteClient(cfg);
+    final String commitTime2 = "002";
+    final String commitTime3 = "003";
+    AtomicReference<Object> writeStatus1 = new AtomicReference<>(null);
+    AtomicReference<Object> writeStatus2 = new AtomicReference<>(null);
+
+    // Create upserts, schedule cleaning, schedule compaction in parallel
+    Future future1 = executor.submit(() -> {
+      final int numRecords = 100;
+      assertDoesNotThrow(() -> {
+        writeStatus1.set(createCommitWithInserts(cfg, client1, "001", commitTime2, numRecords, false));
+      });
+    });
+    // Create upserts, schedule cleaning, schedule compaction in parallel
+    Future future2 = executor.submit(() -> {
+      final int numRecords = 100;
+      assertDoesNotThrow(() -> {
+        writeStatus2.set(createCommitWithInserts(cfg, client2, "001", commitTime3, numRecords, false));
+      });
+    });
+
+    future1.get();
+    future2.get();
+
+    HoodieTableMetaClient tableMetaClient = client.getTableServiceClient().createMetaClient(true);
+
+    // Get inflight instant stream before commit
+    List<HoodieInstant> inflightInstants = client
+            .getTableServiceClient()
+            .getInflightTimelineExcludeCompactionAndClustering(tableMetaClient)
+            .getReverseOrderedInstants()
+            .collect(Collectors.toList());
+
+    // Commit the instants and get instants to rollback in parallel
+    future1 = executor.submit(() -> {
+      long start = System.currentTimeMillis();
+      LOG.info(String.format("Start to commit instant %s", commitTime2));
+      client1.commit(commitTime2, writeStatus1.get());
+      validInstants.add(commitTime2);
+      LOG.info(String.format("commit the instant cost %d ms", System.currentTimeMillis() - start));
+    });
+
+    future2 = executor.submit(() -> {
+      long start = System.currentTimeMillis();
+      LOG.info(String.format("Start to commit instant %s", commitTime3));
+      client2.commit(commitTime3, writeStatus2.get());
+      validInstants.add(commitTime3);
+      LOG.info(String.format("commit the instant %s cost %d ms", commitTime3, System.currentTimeMillis() - start));
+    });
+
+    Future future3 = executor.submit(() -> {
+      // Sleep to let the writer complete the instants
+      assertDoesNotThrow(() -> Thread.sleep(6000));
```

Comment:
we should try to use CountDownLatch to deterministically proceed here.
for eg, lets define a countdownlatch w/ count 2. and count them down at L 624 and 633. And here at L 638, we just need to await for the countdown latch to get to 0 and proceed. also, lets do a timed await and not wait indefinitely.

Reply:
Changed, thanks for your suggestion!  In fact, we can do these operations sequentially(without introducing multi-thread). But I think a multi-thread test can show the case that we want to fix in a better way so we can keep using multi-thread.